File: solutions/Stack/150_evaluate-reverse-polish-notation.py
from typing import List

class Solution:
    def evalRPN(self, tokens: List[str]) -> int:
        stk = []
        for i in tokens:
            if i == "+":
                b, a = stk.pop(), stk.pop()
                stk.append(a + b)

            elif i == "-":
                b, a = stk.pop(), stk.pop()
                stk.append(a - b)

            elif i == "*":
                b, a = stk.pop(), stk.pop()
                stk.append(a * b)

            elif i == "/":
                b, a = stk.pop(), stk.pop()
                stk.append(int(a / b))

            else:
                stk.append(int(i))

        return stk[0]
    
# Time Complexity: O(n)
# Space Complexity: O(n)

File: C++/Stack/150_evaluate-reverse-polish-notation.cpp
#include <iostream>
#include <stack>
#include <vector>

using namespace std;

class Solution {
public:
    int evalRPN(vector<string>& tokens) {
        stack<int> stk;
        int a, b;

        for (string ch: tokens) {
            if (ch == "+") {
                b = stk.top(); stk.pop();
                a = stk.top(); stk.pop();
                stk.push(a + b);
            } else if (ch == "-") {
                b = stk.top(); stk.pop();
                a = stk.top(); stk.pop();
                stk.push(a - b);
            } else if (ch == "*") {
                b = stk.top(); stk.pop();
                a = stk.top(); stk.pop();
                stk.push(a * b);
            } else if (ch == "/") {
                b = stk.top(); stk.pop();
                a = stk.top(); stk.pop();
                stk.push(a / b);
            } else {
                stk.push(stoi(ch));
            }
        }

        return stk.top();
    }
};